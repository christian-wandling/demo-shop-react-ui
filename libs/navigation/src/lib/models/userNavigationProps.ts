import { UserResponse } from '@demo-shop-react-ui/api';

export interface UserNavigationProps {
  user: UserResponse | null;
  className: string;
  'data-testid'?: string;
  onLogin: () => void;
  onLogout: () => void;
  onRegister: () => void;
}
