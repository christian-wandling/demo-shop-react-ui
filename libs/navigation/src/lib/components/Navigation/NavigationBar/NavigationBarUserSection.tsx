import { UserNavigationProps } from '../../../models/userNavigationProps';

export const NavigationBarUserSection = ({ user, onLogin, onLogout, onRegister }: UserNavigationProps) => {
  if (user) {
    return (
      <div className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6">
        <a className="text-sm font-medium text-gray-700  cursor-pointer">
          {user.firstname}&nbsp;{user.lastname}
        </a>
        <span className="h-6 w-px bg-gray-200" aria-hidden="true"></span>
        <a onClick={onLogout} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
          Sign out
        </a>
      </div>
    );
  }

  return (
    <div className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6">
      <a onClick={onLogin} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
        Sign in
      </a>
      <span className="h-6 w-px bg-gray-200" aria-hidden="true"></span>
      <a onClick={onRegister} className="text-sm font-medium text-gray-700 hover:text-indigo-500 cursor-pointer">
        Register
      </a>
    </div>
  );
};
