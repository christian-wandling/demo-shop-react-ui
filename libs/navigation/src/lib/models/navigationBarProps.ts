import { NavigationItem } from './navigationItem';
import { UserResponse } from '@demo-shop-react-ui/api';

export interface NavigationBarProps {
  items: NavigationItem[];
  selectedItem: string;
  user?: UserResponse;
  onOpenSideNavigation: () => void;
  onSelectedItemChange: (item: string) => void;
  onLogin: () => void;
  onRegister: () => void;
  onLogout: () => void;
}
