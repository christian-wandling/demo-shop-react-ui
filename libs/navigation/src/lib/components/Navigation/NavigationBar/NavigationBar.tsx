import { Link } from 'react-router';
import { ProductSearch } from '@demo-shop-react-ui/product';
import { CartIcon } from '@demo-shop-react-ui/shopping';
import { NavigationBarProps } from '../../../models/navigationBarProps';
import { RouteItem } from '../../../models/routeItem';
import { UserNavigation } from '../UserNavigation/UserNavigation';

/**
 * NavigationDesktop renders the desktop version of the navigation.
 * It includes the main navigation bar with logo, links, search and cart.
 */
export const NavigationBar = ({
  items,
  selectedItem,
  user,
  onOpenSideNavigation,
  onSelectedItemChange,
  onLogin,
  onRegister,
  onLogout,
}: NavigationBarProps) => {
  return (
    <header className="sticky top-0 z-20 bg-white" data-testid="navigation-bar">
      <nav aria-label="Top" className="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8">
        <div className="border-b border-gray-200">
          <div className="flex h-16 items-center">
            <button
              onClick={onOpenSideNavigation}
              type="button"
              className="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
              data-testid="navigation-slide-over-toggle">
              <span className="absolute -inset-0.5"></span>
              <span className="sr-only">Open menu</span>
              <svg
                className="h-6 w-6"
                fill="none"
                viewBox="0 0 24 24"
                strokeWidth="1.5"
                stroke="currentColor"
                aria-hidden="true">
                <path strokeLinecap="round" strokeLinejoin="round" d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5" />
              </svg>
            </button>

            <div className="ml-4 flex lg:ml-0 h-8 w-8 relative" data-testid="navigation-bar-logo">
              <Link to="/">
                <span className="sr-only">Demo Shop</span>
                <img className="object-contain" src="/icons/demo-shop.png" alt="Demo shop logo" />
              </Link>
            </div>

            <div className="hidden lg:ml-8 lg:block lg:self-stretch">
              <div className="flex h-full space-x-8">
                {items.map((item, index) => (
                  <div key={index} className="menu-item flex h-full space-x-8">
                    <Link
                      onClick={() => onSelectedItemChange(item.label)}
                      to={(item as RouteItem).options?.route}
                      className={
                        (selectedItem === item.label ? 'text-indigo-600' : 'text-gray-700') +
                        'menu-item-button flex items-center text-sm font-medium hover:text-indigo-500'
                      }
                      data-testid={`navigation-bar-item-${item.label}`}>
                      {item.label.charAt(0).toUpperCase() + item.label.slice(1)}
                    </Link>
                  </div>
                ))}
              </div>
            </div>

            <div className="ml-auto flex items-center">
              <UserNavigation
                className="hidden lg:flex lg:flex-1 lg:items-center lg:justify-end lg:space-x-6"
                data-testid="navigation-bar-user-section"
                user={user}
                onLogin={onLogin}
                onLogout={onLogout}
                onRegister={onRegister}
              />
              <ProductSearch />
              <CartIcon />
            </div>
          </div>
        </div>
      </nav>
    </header>
  );
};
