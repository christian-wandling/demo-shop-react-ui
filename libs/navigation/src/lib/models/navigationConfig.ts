import { NavigationItem } from './navigationItem';

export interface NavigationConfig {
  items: NavigationItem[];
}
