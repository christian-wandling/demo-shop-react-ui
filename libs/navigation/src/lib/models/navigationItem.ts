import { PermissionStrategy } from '@demo-shop/auth';
import { NavigationItemType } from '../enums/navigationItemType';
import { RouteItem } from './routeItem';

/**
 * Base abstract class representing a navigation item in the application
 */
export abstract class NavigationItem {
  /**
   * The type of navigation item
   */
  readonly type!: NavigationItemType;

  /**
   * Creates a new navigation item
   * @param {string} label - Display text for the navigation item
   * @param {number} order - Numeric value determining the display order of the item
   * @param {object} [options] - Optional configuration options
   * @param {PermissionStrategy} [options.permissionStrategy] - Strategy for determining user permissions for this item
   * @param {RouteItem[]} [subItems] - Child navigation items nested under this item
   */
  protected constructor(
    public label: string,
    public order: number,
    public options?: {
      permissionStrategy?: PermissionStrategy;
    },
    public subItems?: RouteItem[]
  ) {}
}
