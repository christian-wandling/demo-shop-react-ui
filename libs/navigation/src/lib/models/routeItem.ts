import { NavigationItemType } from '../enums/navigationItemType';
import { NavigationItem } from './navigationItem';
import { PermissionStrategy } from '@demo-shop-react-ui/auth';

/**
 * Represents a navigation item that links to an Angular route
 */
export class RouteItem extends NavigationItem {
  /**
   * The type of navigation item, always set to ROUTE for RouteItem
   */
  override readonly type: NavigationItemType = NavigationItemType.ROUTE;

  constructor(
    label: string,
    order: number,
    public override options: {
      route: string;
      permissionStrategy?: PermissionStrategy;
      query?: string;
    },
    subItems?: RouteItem[]
  ) {
    super(label, order, options, subItems);
  }
}
