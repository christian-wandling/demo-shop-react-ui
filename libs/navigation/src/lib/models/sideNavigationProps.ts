import { NavigationItem } from './navigationItem';
import { UserResponse } from '@demo-shop-react-ui/api';

export interface SideNavigationProps {
  items: NavigationItem[];
  selectedItem: string;
  user: UserResponse | null;
  isOpen: boolean;
  onSetOpen: (isOpen: boolean) => void;
  onSelectedItemChange: (item: string) => void;
  onLogin: () => void;
  onRegister: () => void;
  onLogout: () => void;
}
