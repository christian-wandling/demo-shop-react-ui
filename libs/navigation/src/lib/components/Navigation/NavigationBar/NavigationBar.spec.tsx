import { beforeEach, describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { NavigationBar } from './NavigationBar';
import { MemoryRouter } from 'react-router';
import { UserNavigationProps } from '../../../models/userNavigationProps';
import { RouteItem } from '../../../models/routeItem';
import { NavigationBarProps } from '../../../models/navigationBarProps';

vi.mock('@demo-shop-react-ui/product', () => ({
  ProductSearch: () => <div data-testid="product-search">Product Search</div>,
}));

vi.mock('../UserNavigation/UserNavigation', () => ({
  UserNavigation: ({ user }: UserNavigationProps) => (
    <div data-testid="navigation-bar-user-section">{user ? 'Logged in' : 'Not logged in'}</div>
  ),
}));

describe('NavigationBar', () => {
  const mockProps: NavigationBarProps = {
    user: {
      email: 'test@example.com',
      id: 0,
      firstname: 'firstname',
      lastname: 'lastname',
      phone: null,
    },
    items: [new RouteItem('products', 101, { route: 'products' }), new RouteItem('orders', 102, { route: 'orders' })],
    selectedItem: 'products',
    onOpenSideNavigation: vi.fn(),
    onSelectedItemChange: vi.fn(),
    onLogin: vi.fn(),
    onRegister: vi.fn(),
    onLogout: vi.fn(),
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('displays the correct navigation items', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    expect(screen.getByText('Products')).toBeTruthy();
    expect(screen.getByText('Orders')).toBeTruthy();
  });

  it('highlights the selected item', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    const homeLink = screen.getByText('Products');
    const productsLink = screen.getByText('Orders');

    expect(homeLink.className).toContain('text-indigo-600');
    expect(productsLink.className).toContain('text-gray-700');
  });

  it('calls onSelectedItemChange when a navigation item is clicked', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    fireEvent.click(screen.getByText('Products'));
    expect(mockProps.onSelectedItemChange).toHaveBeenCalledWith('products');
  });

  it('calls onOpenSideNavigation when the hamburger button is clicked', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    const hamburgerButton = screen.getByRole('button', { name: /open menu/i });
    fireEvent.click(hamburgerButton);
    expect(mockProps.onOpenSideNavigation).toHaveBeenCalled();
  });

  it('renders the logo with correct link', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    expect(screen.getByAltText('Demo shop logo')).toBeTruthy();
  });

  it('renders ProductSearch component', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('product-search')).toBeTruthy();
  });

  it('renders NavigationBarUserSection with correct props', () => {
    render(
      <MemoryRouter>
        <NavigationBar {...mockProps} />
      </MemoryRouter>
    );

    expect(screen.getByTestId('navigation-bar-user-section')).toBeTruthy();
  });

  describe('Snapshots', () => {
    it('matches snapshot', () => {
      const { container } = render(
        <MemoryRouter>
          <NavigationBar {...mockProps} />
        </MemoryRouter>
      );

      expect(container).toMatchSnapshot();
    });
  });
});
