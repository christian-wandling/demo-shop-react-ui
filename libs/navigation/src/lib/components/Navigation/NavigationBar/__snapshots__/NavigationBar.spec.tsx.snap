// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NavigationBar > Snapshots > matches snapshot 1`] = `
<div>
  <header
    class="relative bg-white"
  >
    <nav
      aria-label="Top"
      class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-gray-200"
      >
        <div
          class="flex h-16 items-center"
        >
          <button
            class="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
            type="button"
          >
            <span
              class="absolute -inset-0.5"
            />
            <span
              class="sr-only"
            >
              Open menu
            </span>
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="ml-4 flex lg:ml-0 h-8 w-8 relative"
          >
            <a
              data-discover="true"
              href="/"
            >
              <span
                class="sr-only"
              >
                Demo Shop
              </span>
              <img
                alt="Demo shop logo"
                class="object-contain"
                src="/icons/demo-shop.png"
              />
            </a>
          </div>
          <div
            class="hidden lg:ml-8 lg:block lg:self-stretch"
          >
            <div
              class="flex h-full space-x-8"
            >
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="text-indigo-600menu-item-button flex items-center text-sm font-medium hover:text-indigo-500"
                  data-discover="true"
                  href="/products"
                >
                  Products
                </a>
              </div>
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="text-gray-700menu-item-button flex items-center text-sm font-medium hover:text-indigo-500"
                  data-discover="true"
                  href="/orders"
                >
                  Orders
                </a>
              </div>
            </div>
          </div>
          <div
            class="ml-auto flex items-center"
          >
            <div
              data-testid="user-section"
            >
              User Section
              <button
                data-testid="login-button"
              >
                Login
              </button>
              <button
                data-testid="logout-button"
              >
                Logout
              </button>
              <button
                data-testid="register-button"
              >
                Register
              </button>
            </div>
            <div
              data-testid="product-search"
            >
              Product Search
            </div>
          </div>
        </div>
      </div>
    </nav>
  </header>
</div>
`;

exports[`NavigationBar > renders correctly 1`] = `
<div>
  <header
    class="relative bg-white"
  >
    <nav
      aria-label="Top"
      class="mx-auto max-w-7xl px-4 sm:px-6 lg:px-8"
    >
      <div
        class="border-b border-gray-200"
      >
        <div
          class="flex h-16 items-center"
        >
          <button
            class="relative rounded-md bg-white p-2 text-gray-400 lg:hidden"
            type="button"
          >
            <span
              class="absolute -inset-0.5"
            />
            <span
              class="sr-only"
            >
              Open menu
            </span>
            <svg
              aria-hidden="true"
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              stroke-width="1.5"
              viewBox="0 0 24 24"
            >
              <path
                d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              />
            </svg>
          </button>
          <div
            class="ml-4 flex lg:ml-0 h-8 w-8 relative"
          >
            <a
              data-discover="true"
              href="/"
            >
              <span
                class="sr-only"
              >
                Demo Shop
              </span>
              <img
                alt="Demo shop logo"
                class="object-contain"
                src="/icons/demo-shop.png"
              />
            </a>
          </div>
          <div
            class="hidden lg:ml-8 lg:block lg:self-stretch"
          >
            <div
              class="flex h-full space-x-8"
            >
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="text-indigo-600menu-item-button flex items-center text-sm font-medium hover:text-indigo-500"
                  data-discover="true"
                  href="/products"
                >
                  Products
                </a>
              </div>
              <div
                class="menu-item flex h-full space-x-8"
              >
                <a
                  class="text-gray-700menu-item-button flex items-center text-sm font-medium hover:text-indigo-500"
                  data-discover="true"
                  href="/orders"
                >
                  Orders
                </a>
              </div>
            </div>
          </div>
          <div
            class="ml-auto flex items-center"
          >
            <div
              data-testid="user-section"
            >
              User Section
              <button
                data-testid="login-button"
              >
                Login
              </button>
              <button
                data-testid="logout-button"
              >
                Logout
              </button>
              <button
                data-testid="register-button"
              >
                Register
              </button>
            </div>
            <div
              data-testid="product-search"
            >
              Product Search
            </div>
          </div>
        </div>
      </div>
    </nav>
  </header>
</div>
`;
